src/rust/ide/lib/view/src/graph_editor.rs

```diff
@@ -960,10 +960,69 @@ impl GraphEditorModelWithNetwork {
         node_id
     }
 
+    fn is_node_connected_at_input(&self, node_id:NodeId, crumbs:span_tree::Crumbs) -> bool {
+        if let Some(node) = self.nodes.get_cloned(&node_id) {
+            for in_edge_id in node.in_edges.raw.borrow().iter() {
+                if let Some(edge) = self.edges.get_cloned(in_edge_id) {
+                    if let Some(target) = edge.target() {
+                        if target.node_id == node_id && target.port.as_ref() == &crumbs {
+                            return true
+                        }
+                    }
+                }
+            }
+        }
+        false
+    }
 
     pub fn get_node_position(&self, node_id:NodeId) -> Option<Vector3<f32>> {
         self.nodes.get_cloned_ref(&node_id).map(|node| node.position())
     }
+
+    fn create_edge
+    (&self, edge_click:&frp::Source<EdgeId>,edge_over:&frp::Source<EdgeId>,
+     edge_out:&frp::Source<EdgeId>) -> EdgeId {
+        let edge    = Edge::new(component::Edge::new(&self.scene));
+        let edge_id = edge.id();
+        self.add_child(&edge);
+        self.edges.insert(edge.clone_ref());
+
+        let network = &self.network;
+
+        // TODO implement abstraction for mouse events on components, also with cursor pos.
+        frp::extend! { network
+            eval_ edge.view.frp.shape_events.mouse_down ( edge_click.emit(edge_id));
+            eval_ edge.view.frp.shape_events.mouse_over ( edge_over.emit(edge_id));
+            eval_ edge.view.frp.shape_events.mouse_out ( edge_out.emit(edge_id));
+        }
+
+        edge_id
+    }
+
+    fn new_edge_from_output
+    (&self,edge_click:&frp::Source<EdgeId>,edge_over:&frp::Source<EdgeId>,
+     edge_out:&frp::Source<EdgeId>) -> EdgeId {
+        let edge_id = self.create_edge(edge_click,edge_over,edge_out);
+        let first_detached = self.edges.detached_target.is_empty();
+        self.edges.detached_target.insert(edge_id);
+        if first_detached {
+            self.frp.some_edge_targets_detached.emit(());
+        }
+        edge_id
+    }
+
+    fn new_edge_from_input
+     (&self,edge_click:&frp::Source<EdgeId>,edge_over:&frp::Source<EdgeId>,
+      edge_out:&frp::Source<EdgeId>) -> EdgeId {
+        let edge_id = self.create_edge(edge_click,edge_over,edge_out);
+        let first_detached = self.edges.detached_source.is_empty();
```

Comment:
alignment -- in other places as well, please make a fast self-review on github and check for all such places.